src/internal/m365/graph/concurrency_middleware.go

```diff
@@ -20,34 +20,41 @@ import (
 // "how many calls at one time"
 // ---------------------------------------------------------------------------
 
-// concurrencyLimiter middleware limits the number of concurrent requests to graph API
-type concurrencyLimiter struct {
+// concurrencyLimiterMiddleware middleware limits the number of concurrent requests to graph API
+type concurrencyLimiterMiddleware struct {
 	semaphore chan struct{}
 }
 
 var (
-	once                  sync.Once
-	concurrencyLim        *concurrencyLimiter
-	maxConcurrentRequests = 4
+	once                                sync.Once
+	concurrencyLimitMiddlewareSingleton = &concurrencyLimiterMiddleware{}
+	maxConcurrentRequests               = 4
 )
 
-func generateConcurrencyLimiter(capacity int) *concurrencyLimiter {
+func generateConcurrencyLimiter(capacity int) *concurrencyLimiterMiddleware {
```

Comment:
this function should be reworked to return just the channel. I don't see a reason given the code we have right now to return a full middleware and then pull out one of the internal fields of it while discarding the rest

Reply:
I'm not against the rework, but it does give us a weird sort of regression protection in this state.  Since we initialize a non-nil middleware struct holding a nil channel, we can signal for an unconfigured (trinary) state to catch occasions where we forget or slice out initialization of the middleware.  An  absent signal is what led to this bug in the first place.